File: Perfect Sum Problem - GFG/perfect-sum-problem.js
//{ Driver Code Starts
//Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let [n,sum] = readLine().trim().split(" ").map((x) => parseInt(x));
        let arr = readLine().trim().split(" ").map((x) => parseInt(x));
        let obj = new Solution();
        let res = obj.perfectSum(arr,n,sum);
        if(res === -0)
            res = 0;
        console.log(res);
        
    }
}
// } Driver Code Ends


//User function Template for javascript

/**
 * @param {number[]} arr
 * @param {number} n
 * @param {number} sum
 * @return {number}
*/

class Solution {
    constructor() {
        this.mod = 1000000007;
        this.dp = [];
        this.N = 0;
    }
    
    fun(pos, sum, arr) {
        if (pos >= this.N)
            return sum === 0 ? 1 : 0;
            
        if (this.dp[pos][sum] !== -1)
            return this.dp[pos][sum];
            
        this.dp[pos][sum] = 0;

        this.dp[pos][sum] = (this.dp[pos][sum] + this.fun(pos + 1, sum, arr)) % this.mod;

        if (arr[pos] <= sum)
            this.dp[pos][sum] = (this.dp[pos][sum] + this.fun(pos + 1, sum - arr[pos], arr)) % this.mod;

        return this.dp[pos][sum];
    }

    perfectSum(arr, n, sum) {
        this.dp = new Array(n + 1).fill().map(() => new Array(sum + 2).fill(-1));
        this.N = n;

        return this.fun(0, sum, arr);
    }
}

File: Perfect Sum Problem - GFG/perfect-sum-problem.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;
class GfG
{
    public static void main(String args[])
        {
            Scanner sc = new Scanner(System.in);
            int t = sc.nextInt();
            while(t-->0)
                {
                    int n = sc.nextInt();
                    int sum = sc.nextInt();
                    int arr[] = new int[n];
                    for(int i = 0;i<n;i++)
                    arr[i] = sc.nextInt();
                    
                    Solution ob = new Solution();
                    System.out.println(ob.perfectSum(arr,n,sum));
                }
        }
}    
// } Driver Code Ends


class Solution{

    int mod=1000000007; // Store the value of 10^9 + 7 for later use
    int dp[][]; // Create a 2D array for memoization
    
	public int perfectSum(int arr[],int n, int sum) 
	{ 
	    // Your code goes here
	    dp=new int[n+1][sum+1]; // Initialize the memoization array with size (n+1) x (sum+1)
	    for(int i=0;i<n+1;i++){
	        Arrays.fill(dp[i],-1); // Fill the memoization array with -1 for later check
	    }
	    return fun(arr,0,sum); // Call the recursive helper function
	}
	
	public int fun(int arr[],int i,int sum)
	{
	    if(sum==0){ // If the desired sum is 0
	        int j=i;
	        int p=1; // Initialize a variable p with 1
	        while(j<arr.length && arr[j]==0){ // Count the number of 0's in the array starting from index i
	            p++; // Increment the variable p
	            j++; // Move to the next index
	        }
	        return dp[i][sum]=p; // Store the value of p in the memoization array and return p
	    }
	    if(i==arr.length){ // If all array elements are processed
	        return dp[i][sum]=0; // Return 0
	    }
	    if(dp[i][sum]!=-1)return dp[i][sum]; // If the value is already computed, return it from the memoization array
	    if(arr[i]<=sum) // If the current element is less than or equal to the desired sum
	    return dp[i][sum]=(fun(arr,i+1,sum)+fun(arr,i+1,sum-arr[i]))%mod; // Recursively compute the sum with and without the current element and store it in the memoization array
	    else
	    return dp[i][sum]=fun(arr,i+1,sum); // If the current element is greater than the desired sum, recursively compute the sum without the current element and store it in the memoization array
	}
}